### PR TITLE
Unistore: Add GetFolders endpoint backed by UnifiedStorage

### PR DESCRIPTION
**Pull request by filewalkwithme**
**What is this feature?**

This PR adds an implemention of `GET /api/folders` endpoint backed by Unified Storage.
This endpoints is responsible to List Folders, given a parent UID.

**Why do we need this feature?**

This is required as part of the efforts of onboarding Folders into the Unified Storage

**Who is this feature for?**

@grafana/grafana-search-and-storage 

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/search-and-storage-team/issues/131

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: pkg/tests/apis/folder/folders_test.go [hunk 1]
```diff
@@ -926,3 +926,134 @@ func testDescription(description string, expectedErr error) string {
...
+			expectedCode: http.StatusOK,
+			expectedOutput: []dtos.FolderSearchHit{
+				dtos.FolderSearchHit{UID: "bar", Title: "Folder 2", ParentUID: "foo"},
+			},
+			permissions: folderCreatePermission,
+		},
+	}
+
+	// test on all dualwriter modes
+	for mode := 1; mode <= 4; mode++ {
```
Comment (suntala) [inline pkg/tests/apis/folder/folders_test.go:990]: Nice!

File: pkg/tests/apis/folder/folders_test.go [hunk 2]
```diff
@@ -926,3 +926,134 @@ func testDescription(description string, expectedErr error) string {
...
+			ResourceAttribute: "uid",
+			ResourceID:        "*",
+		},
+	}
+
+	folder1 := "{ \"uid\": \"foo\", \"title\": \"Folder 1\"}"
+	folder2 := "{ \"uid\": \"bar\", \"title\": \"Folder 2\", \"parentUid\": \"foo\"}"
+	folder3 := "{ \"uid\": \"qux\", \"title\": \"Folder 3\"}"
+
+	tcs := []testCase{
```
Comment (suntala) [inline pkg/tests/apis/folder/folders_test.go:958]: Looks like we are missing a test case for this: https://github.com/grafana/grafana/blob/main/pkg/api/folder_test.go#L452

"get folder by UID should return parent folders redacted if nested folder are enabled and user does not have read access to parent folders"

Comment (filewalkwithme) [reply pkg/tests/apis/folder/folders_test.go:958]: Actually this test case refers to the `GetFolderByUID` endpoint, not the `GetFolders` endpoint which is the subject of this PR.

In fact, I realized now that there are no tests for the legacy `GetFolders` endpoijnt.
I am going to remove the comment that says that this new test is a counterpart of legacy.

Let me know if this makes sense.

Comment (suntala) [reply pkg/tests/apis/folder/folders_test.go:958]: Makes a lot of sense. I only see tests for the happy path though. Is there anything we can add relating to permissions? For instance, when would we expect non-200 status codes? For example, one scenario I can think of testing is the following. What happens to the list response if there's an issue with one of the folders (for example converting it to legacy folder DTO)?

Comment (filewalkwithme) [reply pkg/tests/apis/folder/folders_test.go:959]: Hey @suntala , I have added two new testcases:
- listing folders of a parent that does not exists: 404-Not Found
- listing folders from another org: 403-Forbidden

File: pkg/registry/apis/folders/register.go [hunk 3]
```diff
@@ -189,8 +189,10 @@ func (b *FolderAPIBuilder) GetAuthorizer() authorizer.Authorizer {
 func authorizerFunc(ctx context.Context, attr authorizer.Attributes) (*authorizerParams, error) {
 	verb := attr.GetVerb()
 	name := attr.GetName()
-	if (!attr.IsResourceRequest()) || (name == "" && verb != utils.VerbCreate) {
-		return nil, errNoResource
+	if verb != utils.VerbList {
```
Comment (suntala) [inline pkg/registry/apis/folders/register.go:192]: Wondering if we should still be checking that it's a resource request here:

```suggestion
    if (!attr.IsResourceRequest()) || (verb != utils.VerbList) {
```

**Review by suntala — DISMISSED** [review]
Nice updates 👍  Just one ~~non-blocking~~ suggestion.

File: pkg/registry/apis/folders/register.go [hunk 4]
```diff
@@ -189,8 +189,10 @@ func (b *FolderAPIBuilder) GetAuthorizer() authorizer.Authorizer {
...
+		if (!attr.IsResourceRequest()) || (name == "" && verb != utils.VerbCreate) {
+			return nil, errNoResource
+		}
 	}
```
Comment (suntala) [inline pkg/registry/apis/folders/register.go:196]: I assume we should still be checking that it's a resource request here:

```suggestion
	if (!attr.IsResourceRequest()) || (name == "" && verb != utils.VerbCreate && verb != utils.VerbList) {
		return nil, errNoResource
	}
```

**Comment by filewalkwithme** [comment]
@suntala Fair enough, I have accepted your suggestion

Could you take another look?

**Review by suntala — APPROVED** [review]
🚀

File: pkg/tests/apis/folder/folders_test.go (hunk 1 above)
Comment (leonorfmartins) [reply pkg/tests/apis/folder/folders_test.go:1016]: 🤔  why dont use this [pattern](https://github.com/grafana/grafana/blob/main/pkg/tests/apis/folder/folders_test.go#L127) when writing the tests? Doesn't it get confusing?

Comment (leonorfmartins) [reply pkg/tests/apis/folder/folders_test.go:1016]: I had actually created this [issue](https://github.com/grafana/grafana-org/issues/271) to cover all dual writer modes like [playlists](https://github.com/grafana/grafana/blob/main/pkg/tests/apis/playlist/playlist_test.go) is doing (and how we started to write folder tests in this file).

File: pkg/registry/apis/folders/register.go [hunk 5]
```diff
@@ -189,7 +189,7 @@ func (b *FolderAPIBuilder) GetAuthorizer() authorizer.Authorizer {
 func authorizerFunc(ctx context.Context, attr authorizer.Attributes) (*authorizerParams, error) {
 	verb := attr.GetVerb()
 	name := attr.GetName()
-	if (!attr.IsResourceRequest()) || (name == "" && verb != utils.VerbCreate) {
+	if (!attr.IsResourceRequest()) || (name == "" && verb != utils.VerbCreate && verb != utils.VerbList) {
```
Comment (leonorfmartins) [inline pkg/registry/apis/folders/register.go:192]: We have some tests for the authoriser that we can extend here to include list use case: https://github.com/grafana/grafana/blob/main/pkg/registry/apis/folders/register_test.go#L25

File: pkg/tests/apis/folder/folders_test.go (hunk 1 above)
Comment (filewalkwithme) [reply pkg/tests/apis/folder/folders_test.go:1016]: > 🤔 why dont use this [pattern](https://github.com/grafana/grafana/blob/main/pkg/tests/apis/folder/folders_test.go#L127) when writing the tests? Doesn't it get confusing?

Hi @leonorfmartins,  the [pattern](https://github.com/grafana/grafana/blob/b0fe898fa1db3dd6524517741146c9c220b0023f/pkg/tests/apis/folder/folders_test.go#L127) you mentioned is testing the requests to the API Server directly. On the other hand `TestFoldersGetAPIEndpointK8S`, which is being introduced by this PR cares about testing the k8s version of the `GET /api/folders` endpoint. 

They are not testing the same thing.

On the other hand, your comment made me realize that we need to extend `TestIntegrationFoldersApp` to include CRUD tests for Listing folders.

File: pkg/registry/apis/folders/register.go (hunk 5 above)
Comment (filewalkwithme) [reply pkg/registry/apis/folders/register.go:192]: That's true! I added some tests for List Folders there!

File: pkg/tests/apis/folder/folders_test.go (hunk 1 above)
Comment (filewalkwithme) [reply pkg/tests/apis/folder/folders_test.go:1026]: @leonorfmartins I added a test for List Folders to `TestIntegrationFoldersApp`

**Comment by filewalkwithme** [comment]
@leonorfmartins I have added testcases for the Authorizer and also for `TestIntegrationFoldersApp`.

Could you take another look?

**Review by leonorfmartins — APPROVED** [review]

**Pull request merged**